pyterrier/apply.py

```diff
@@ -175,24 +179,49 @@ def generic(fn : Callable[[pd.DataFrame], pd.DataFrame], *args, batch_size=None,
         Arguments:
             fn(Callable): the function to apply to each row
             batch_size(int or None): whether to apply fn on batches of rows or all that are received
-            verbose(bool): Whether to display a progress bar over batches (only used if batch_size is set).
-
-        Example::
+            verbose(bool): Whether to display a progress bar over batches (only used if batch_size is set, and iter is not set).
+            iter(bool): Whether to use the iter-dict API - if-so, then ``fn`` receives an iterable, and returns an iterable. 
 
-            # this transformer will remove all documents at rank greater than 2.
+        Example (dataframe)::
 
             # this pipeline would remove all but the first two documents from a result set
             pipe = pt.terrier.Retriever(index) >> pt.apply.generic(lambda res : res[res["rank"] < 2])
 
+         Example (iter-dict)::
+
+            # this pipeline would simlarly remove all but the first two documents from a result set
+            def _fn(iterdict):
+                for result in iterdict:
+                    if result["rank"] < 2:
+                        yield result
+            pipe1 = pt.terrier.Retriever(index) >> pt.apply.generic(_fn, iter=True)
+
+            # transform_iter() can also return an iterable, so returning a list is also permissible
+            pipe2 = pt.terrier.Retriever(index) >> pt.apply.generic(lambda res: [row for row in res if row["rank"] < 2], iter=True)
+
     """
+    if iter:
```

Comment:
Needing an `iter` argument here is a bit unfortunate since it doesn't work the same way as the other applies.

Reply:
its common across all that take more than one row as input...
a Series is close to a Dict
a DataFrame is not close to an IterDict

Reply:
Yeah, I understand the necessity. The difference is something that might confuse users though.

Reply:
i think the extra functionality is hidden to most users, e.g. through iter=True additional kwarg. The overview table in apply.rst is a bit messy now though.